File: leetcode/algorithms/array/move-zeroes.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Move Zeroes"
*/

/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function(nums) {
    let i=0;
    let zeroStringLength=0;
    while(i<nums.length){
        if(nums[i]===0){
            zeroStringLength++;
            i++;
        } else {
            if(zeroStringLength){
                nums.push(...nums.splice(i-zeroStringLength,zeroStringLength));
                i=i-zeroStringLength;
                zeroStringLength=0;
            } else {
                i++;
            }
        }
    }
};

File: leetcode/algorithms/two-ptr/move-zeroes.java
class Solution {
    public void moveZeroes(int[] nums) {
        int zi = 0,ni = 0;
        while(zi<nums.length){
           if(nums[zi]==0){
               ni = zi;
               while(ni<nums.length && nums[ni]==0){
                   ++ni;
               }
                if(ni==nums.length){
                    break;
                }
                   
                int temp = nums[zi];
                nums[zi] = nums[ni];
                nums[ni] = temp;
           }
            zi++;
        }
    }
}